### PR TITLE
Plug-in: Skilling Calculator

### PR DESCRIPTION
**Pull request by Kruithne**
Hello.

I'll preface by saying that I only found RuneLite yesterday, so I'm not fully familiar with how things work on this project, but as far as I know I've followed the coding guidelines and made proper use of the API. If there are any mistakes, please let me know!

This PR adds a Skilling Calculator plug-in, which as the name suggests allows players to calculate how many actions are needed to reach a specific XP goal.

runelite_2018-03-12_19-28-35

Certain bonuses, such as skilling sets, can be toggled on for some of the calculators where applicable, which adjust the calculations accordingly.

Thanks again.

### CODE REVIEW

**Comment by Adam-** [comment]
I don't see why this couldn't go in once we get a better way to get item images. You may be able to use the ItemClient to fetch them from the rs ge api, but it generally only contains items tradeable on the ge. Obviously putting all of the images in the tree is not acceptable.

**Comment by Kruithne** [comment]
As you requested, I have moved away from using local resources as much as possible. I've managed to get the `ItemManager` class to work inside the plug-in, and now rely on that for 852 out of 1020 of the icons. This reduces the amount of local resource icons to only 168 (36.5kb) for icons that I could not get from items.

Icons that have not moved over to this new system include NPC heads, spell sprites, and hunting creatures. If a method can be suggested to access these sprites easily, I will swap those over too, but I was unable to find any correlating items.

While this has reduced the amount of local resources as you requested, it has introduced a strange bug which I can only assume lies inside the `ItemManager` or deeper. Some sprites are occasionally rendered with weird textures behind/over them.

runelite_2018-03-13_15-41-54

File: runelite-client/src/main/java/net/runelite/client/plugins/skillcalculator/SkillCalculator.java [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+	{
+		return Math.min(MAX_LEVEL, Math.max(MIN_LEVEL, input));
+	}
+
+	private static int EnforceXPBounds(int input)
+	{
+		return Math.min(MAX_XP, Math.max(0, input));
+	}
+
+	private static int SkillLevelToXP(int level)
```
Comment (SomeoneWithAnInternetConnection) [inline runelite-client/src/main/java/net/runelite/client/plugins/skillcalculator/SkillCalculator.java]: I think we've got code for this in the `Experience` class in `net.runelite.api`. If we're missing functionality, it'd probably be best to add it there.

**Comment by ChaoticConundrum** [comment]
This is really cool, very nice work. I'm having a few issues. The big one is game crashes when selecting a skill in the grid. Output below. Presumably there is some unsafe usage of the runescape-api (wrapped by the runelite-api) which is producing this error. Note this is on a branch with master as of 8f604216c merged.
```
Error: p.jj() jk.iq() jk.iq() kq.ix() client.gu() client.ah() bh.h() | la:686 p:11457 jk:9509 jk:9595 kq:9321 client:3533 client:922 bh:369 bh:323 Thread:748 | java.lang.ArrayIndexOutOfBoundsException: 5675
java.lang.ArrayIndexOutOfBoundsException: 5675
	at la.ah(la.java:686)
	at p.jj(p.java:11457)
	at jk.iq(jk.java:9509)
	at jk.iq(jk.java:9595)
	at kq.ix(kq.java:9321)
	at client.gu(client.java:3533)
	at client.ah(client.java:922)
	at bh.h(bh.java:369)
	at bh.run(bh.java:323)
	at java.lang.Thread.run(Thread.java:748)
error_game_crash
```
And another occurrence, different trace.
```
java.lang.ArrayIndexOutOfBoundsException: 524523
	at eu.c(eu.java:552)
	at eu.o(eu.java:470)
	at ew.ad(ew.java:1308)
	at ew.ab(ew.java:923)
	at ew.copy$drawRegion(ew.java:811)
	at ew.ak(ew.java:53)
	at e.gd(e.java:3878)
	at jk.iq(jk.java:9470)
	at kq.ix(kq.java:9321)
	at client.gu(client.java:3533)
	at client.ah(client.java:922)
	at bh.h(bh.java:369)
	at bh.run(bh.java:323)
	at java.lang.Thread.run(Thread.java:748)
Error: e.gd() jk.iq() kq.ix() client.gu() client.ah() bh.h() | eu:552 eu:470 ew:1308 ew:923 ew:811 ew:53 e:3878 jk:9470 kq:9321 client:3533 client:922 bh:369 bh:323 Thread:748 | java.lang.ArrayIndexOutOfBoundsException: 524523
error_game_crash
```


There is also this less serious exception:
```
Exception in thread "AWT-EventQueue-0" java.lang.IllegalArgumentException: input == null!
	at javax.imageio.ImageIO.read(ImageIO.java:1348)
	at net.runelite.client.plugins.skillcalculator.CacheItemIcon.getLocalIcon(CacheItemIcon.java:59)
	at net.runelite.client.plugins.skillcalculator.UIActionSlot.<init>(UIActionSlot.java:52)
	at net.runelite.client.plugins.skillcalculator.SkillCalculator.renderActionSlots(SkillCalculator.java:134)
	at net.runelite.client.plugins.skillcalculator.SkillCalculator.openCalculator(SkillCalculator.java:97)
	at net.runelite.client.plugins.skillcalculator.SkillCalculatorPanel.openCalculator(SkillCalculatorPanel.java:103)
	at net.runelite.client.plugins.skillcalculator.SkillCalculatorPanel.lambda$addCalculatorButtons$0(SkillCalculatorPanel.java:79)
	at javax.swing.AbstractButton.fireActionPerformed(AbstractButton.java:2022)
	at javax.swing.AbstractButton$Handler.actionPerformed(AbstractButton.java:2348)
	at javax.swing.DefaultButtonModel.fireActionPerformed(DefaultButtonModel.java:402)
	at javax.swing.DefaultButtonModel.setPressed(DefaultButtonModel.java:259)
	at javax.swing.plaf.basic.BasicButtonListener.mouseReleased(BasicButtonListener.java:252)
	at org.pushingpixels.substance.internal.utils.RolloverButtonListener.mouseReleased(RolloverButtonListener.java:126)
	at java.awt.AWTEventMulticaster.mouseReleased(AWTEventMulticaster.java:289)
	...
```

**Comment by ChaoticConundrum** [comment]
I had a quick look at it, note that `ItemManager.getImage` eventually calls directly into RS client rendering code so it is not safe to call it from an AWT thread.

**Comment by LeviSchuck** [comment]
This looks fun.

**Comment by Kruithne** [comment]
This should be ready to merge unless any further changes are wanted.

**Comment by Adam-** [comment]
It may be worth it for someone to look at how hard it is to get the sprites out of the client as a bufferedimage instead of commiting all of these... otherwise overall it seems fine.

**Comment by Kruithne** [comment]
Some visual updates have been added based on feedback on Discord, however @psikoi has offered to sprinkle some redesign magic onto it once it's finished.

runelite_2018-05-06_02-27-27

As shown above, support has been added to "combine" actions together, useful for calculating skilling actions that are commonly done together.

Regarding the built-in sprite count.. it went up with my latest commit (by 55 files) since some files were missing that prevented the PR from working. I fully intend to remove all of the sprites from the PR once @Abextm has finished work on the AsyncBufferedImage class, allowing programmatic access of sprites safely.

**Comment by permedd** [comment]
soo when is this going to be released?

**Comment by permedd** [comment]
is there a way to put a extra box where we can type the name? instead of writing the xp/level etc?

so example if i write a players name it should automatically fill up the boxes for xp/level etc it will be so much easier to use rather then putting the level and the xp etc.

**Comment by Kruithne** [comment]
@permedd The skill calculator is already merged, so it won't be added before it is released; especially not with everything that is going on right now.

It's a good idea though. I've added it to my list of improvements to make to the skill calculator, once everything is cleared up I'll start adding them.

**Comment by permedd** [comment]
thats great man thanks.

**Comment by EliasLahham** [comment]
@Kruithne 

See this image for more context: http://prntscr.com/jk1m7a

In the image I wanted to know all the methods from lv 70-78 construction, but when I do that it lists a million options with no way to filter. If there was a way to type in a search bar and search something like "oak larders" or "mahogany tables" that would make it a bit more organized.

**Comment by Kruithne** [comment]
@EliasLahham It should be organized by the level at which you are able to create each thing, as that's how all other skill calculators function. I can see the use of a filter box, it will be added to the list after the upcoming redesign by @psikoi.

**Comment by EliasLahham** [comment]
@Kruithne Ok thank you!

**Pull request merged**